KEP: Introduce a new timeout to WaitForPodsReady config (#2737)
kubernetes-sigs/kueue: keps/349-all-or-nothing/README.md

- guarantee that two jobs would not schedule pods concurrently. Example
scenarios in which two jobs may still concurrently schedule their pods:

> - when succeeded pods are replaced with new because job's parallelism is less than its completions;

Reviewer: ```suggestion
- when succeeded pods are replaced with new because job's parallelism is less than its completions.
```

My use case can be supported by enabling `waitForPodsReady` in the Kueue
configuration.

#### Story 2

As a Kueue administrator I want to ensure that a Workload will be evicted after

> configured timeout if a pod fails during runtime and can't be scheduled.

Reviewer: ```suggestion
configured timeout if a pod fails during its execution and the replacement Pod can't be scheduled.
```

	// enforced only if waitForPodsReady.enable=true. If unspecified, it defaults to 5min.
	// +optional
	TimeoutSeconds *int64 `json:"timeoutSeconds,omitempty"`

	// replacementTimeoutSeconds defines optional time duration in seconds, relative to
	// pod's failure after a Workload is Admitted and running.
	// After exceeding the timeout the corresponding job gets suspended again
	// and moved to the ClusterQueue's inadmissibleWorkloads list. The timeout is

> 	// enforced only if waitForPodsReady.enable=true. If unspecified, it defaults to 5min.

Reviewer: ```suggestion
	// enforced only if waitForPodsReady.enable=true. If unspecified, it defaults to 5min.
	// +optional
```

> 	ReplacementTimeoutSeconds *int64

Reviewer: The alternative API to consider could be:
```
  afterStart:
     timeout: 5m
```
This could be more flexible if we want more configuration options. 
Also, technically, PodsReady=False does not necessarily require a pod replacement. It could be the same pod, if the Pod has readinessProbe defined, and it was just temporarily returning errors, but it could start responding ok after a while.

I would be leaning towards that more generic API. WDYT @tenzen-y ?

Reviewer: FWIW: I would also be supportive of `AfterStartTimeout`.

Author: Sounds good to me, especially given that we don't always need replacement as you described

Reviewer 2: `FailureRecoveryTimeout`? It's a replacement in some cases, but it could also be a matter of the pod(s) becoming healthy again.

Reviewer 3: I think the key point is how many times we will assume the failure once the Pod is healthy.
Will we assume only 1? or infinity?

Author: >> FailureRecoveryTimeout? It's a replacement in some cases, but it could also be a matter of the pod(s) becoming healthy again.

SGTM

>> I think the key point is how many times we will assume the failure once the Pod is healthy.
Will we assume only 1? or infinity?

Currenztly, `WaitForPodsReady` API contains `RequeuingStrategy` field that defines how many times Kueue should allow pod to retry. One option would be to use that `RequeuingStrategy` configuration for both timeouts (we could also reset `RequeueState` field in Workload's Status so the amount of retries won't be combined). Another alternative would be to add a analogous field for the new timeout so admins could define different strategy for different timeout. Personally I believe the second option is a bit of an overkill, and would make admin's experience worse, but I'm more than happy to hear your opinion on that

Reviewer 4: >I think the key point is how many times we will assume the failure once the Pod is healthy.
Will we assume only 1? or infinity?

Do you ask how many times we allow pods fail? From KEP I see that we allow infinite number of failures.

Reviewer 4: +1 to FailureRecoveryTimeout name

Author: After revisiting naming I suggest `recoveryTimeoutSeconds` as I believe it's rather clear that recovery is because of some failure. I've updated the KEP

WDYT @yaroslava-serdiuk ?

Reviewer 4: recoveryTimeout SGTM as well

Reviewer: I like the new time too

	// enforced only if waitForPodsReady.enable=true. If unspecified, it defaults to 5min.
	// +optional
	TimeoutSeconds *int64 `json:"timeoutSeconds,omitempty"`

> 	// replacementTimeoutSeconds defines optional time duration in seconds, relative to

Reviewer: We need to make sure that when the timeout is not specified, the Job does not get evicted immediately (see comment https://github.com/kubernetes-sigs/kueue/pull/2725#discussion_r1745970954) .
Please add a note about that.

Author: Updated description of the field so the value defaults to 3mins

	// RequeuingStrategy defines the strategy for requeuing a Workload.
	// +optional
	RequeuingStrategy *RequeuingStrategy `json:"requeuingStrategy,omitempty"`

	// recoveryTimeoutSeconds defines optional time duration in seconds, relative to
	// pod's failure expressed by PodsReady=false condition, after a Workload is Admitted and running.
	// After exceeding the timeout the corresponding job gets suspended again
	// and moved to the ClusterQueue's inadmissibleWorkloads list. The timeout is
	// enforced only if waitForPodsReady.enable=true. Defaults to 3 mins.
	// +optional

> 	recoveryTimeoutSeconds *int64 `json:"recoveryTimeoutSeconds,omitempty"`

Reviewer 4: I think it inconsistent to have Timeout *metav1.Duration and recoveryTimeoutSeconds *int64

Author: That's a good point. However, I believe that in general we want to gravitate towards using ints and names with units in it, as above. It's a good practice since it makes API more versatile

Reviewer: Actually I think I would prefer consistency, and human readability, here for global configuration.

 indeed we tend to use units suffix in API exposed to clients, but for configuration API this is mostly manual by admin. 

the reason to avoid the Duration syntax in client facing clients is that some clients in other languages than go may not know the syntax

Author: I see, in that case let's make it consistent with the rest of the API

	// RequeuingStrategy defines the strategy for requeuing a Workload.
	// +optional
	RequeuingStrategy *RequeuingStrategy `json:"requeuingStrategy,omitempty"`

> 	// recoveryTimeout defines optional time duration, relative to
> 	// pod's failure expressed by PodsReady=false condition, after a Workload is Admitted and running.
> 	// After exceeding the timeout the corresponding job gets suspended again
> 	// and moved to the ClusterQueue's inadmissibleWorkloads list. The timeout is
> 	// enforced only if waitForPodsReady.enable=true. Defaults to 3 mins.

Reviewer: ```suggestion
	// recoveryTimeout defines an optional timeout, measured since the
	// last transition to the PodsReady=false condition after a Workload is Admitted and running.
	// Such a transition may happen when a Pod failed and the replacement Pod 
	// is awaited to be scheduled.
	// After exceeding the timeout the corresponding job gets suspended again
	// and requeued. The timeout is enforced only if waitForPodsReady.enable=true. 
	// Defaults to 3 mins.
```

	// RequeuingStrategy defines the strategy for requeuing a Workload.
	// +optional
	RequeuingStrategy *RequeuingStrategy `json:"requeuingStrategy,omitempty"`

	// RecoveryTimeout defines an optional timeout, measured since the
	// last transition to the PodsReady=false condition after a Workload is Admitted and running.
	// Such a transition may happen when a Pod failed and the replacement Pod 
	// is awaited to be scheduled.
	// After exceeding the timeout the corresponding job gets suspended again
	// and requeued after the backoff delay. The timeout is enforced only if waitForPodsReady.enable=true. 

> 	// Defaults to 3 mins.

Reviewer 4: nit: I suggest defaulting to timeout, which is 5 mins

Reviewer 4: disregard, it doesn't make sense and we want to have a low default time in case timeout is set to high number.

We introduce a new workload condition, called `PodsReady`, to indicate
if the workload's startup requirements are satisfied. More precisely, we add

> the condition when `job.status.ready + job.status.uncountedTerimnatedPods + job.status.succeeded` is greater or equal

Reviewer: ```suggestion
the condition when `job.status.ready + job.status.uncountedTerminatedPods + job.status.succeeded` is greater or equal
```

Reviewer: I would also suggest to add a note why we do this somewhere in the KEP update. It could be Design Details or Notes section.

Author: Could you elaborate on that? It is in the Design Details section

Reviewer: I see, the formula is in Design Details, but I would like to have a short note **why** we have `job.status.uncountedTerimnatedPods` here. Maybe something along the lines:
Note that we count `job.status.uncountedTerimnatedPods` - this is meant to prevent flickering of the `PodsReady` condition when pods are transitioning to the Succeeded state.

Author: Done

	// RequeuingStrategy defines the strategy for requeuing a Workload.
	// +optional
	RequeuingStrategy *RequeuingStrategy `json:"requeuingStrategy,omitempty"`

	// RecoveryTimeout defines an optional timeout, measured since the
	// last transition to the PodsReady=false condition after a Workload is Admitted and running.
	// Such a transition may happen when a Pod failed and the replacement Pod 
	// is awaited to be scheduled.
	// After exceeding the timeout the corresponding job gets suspended again
	// and requeued after the backoff delay. The timeout is enforced only if waitForPodsReady.enable=true. 
	// Defaults to 3 mins.
	// +optional

> 	RecoveryTimeout *metav1.Duration `json:"recoveryTimeout,omitempty"`

Reviewer 3: As we discussed in the previous meeting, can we disable this and avoid defaulting with

```yaml
recoveryTimeout: {}
```
?

Reviewer 3: I synced with @PBundyra offline. In conclusion, Patryk found that the above approach is technically challenging.

So, he proposed the following solutions:

> - _ If not specified otherwise, .recoveryTimeout is null, and it doesn't impact Kueue's behavior_
> - _Put the 3mins in Kueue's release manifest, then a user that uncomments and enables it, sees that it's. there_
> - _For existing users that use configmap without .recoveryTimeout :_
>     - _Describe the extension in release notes and mark is as important
>     - _Document it well on WaitForPodsReady website in Kueue's documentation_
>     - _Recommend it in documentation for TAS_

Originally posted by @PBundyra

### Timeout on reaching the PodsReady condition

We introduce two timeouts defined in the `waitForPodsReady.timeoutSeconds` and `waitForPodsReady.replacementTimeoutSeconds` fields.

First one tracks the time between job getting unsuspended (the time of unsuspending a job is marked by the Job's
`job.status.startTime` field) and reaching the `PodsReady=true` condition.

> Second one tracks the time between changing `PodsReady` condition to `false` after the job is running and reaching the

Reviewer: How are we going to know if this is the first `PodsReady=False`? Describe how we can distinguish the two scenarios.

Author: Added, thanks!

### Timeout on reaching the PodsReady condition

We introduce two timeouts defined in the `waitForPodsReady.timeoutSeconds` and `waitForPodsReady.replacementTimeoutSeconds` fields.

First one tracks the time between job getting unsuspended (the time of unsuspending a job is marked by the Job's
`job.status.startTime` field) and reaching the `PodsReady=true` condition.

> Second one tracks the time between changing `PodsReady` condition to `false` after the job is running and reaching the

Reviewer 2: how do you plan to track this time?

Author: I plan to use timestamp in the `PodsReady` condition to calculate how much time passed

Reviewer 5: If we wanted to support additional fault recovery capabilities, it could be desirable to have a separate `PodsUnhealthy` condition on the Workload instead of overloading `PodsReady`.

Author: What semantic differences do you imagine between `PodsUnhealthy`/`WorkloadUnhealthy` and `PodsReady`?

> `PodsReady=true` condition once again.

Reviewer 3: Before I review the API, let me clarify the duration of this new timeout if my understanding is correctly.
IIUC, exiting, and new timeouts indicate the below durations, right?

- `timeoutSeconds`:

```mermaid
stateDiagram-v2
    [*] --> Suspended=false: Workload is admitted
    Suspended=false --> PodsReady=false
    PodsReady=false --> PodsReady=false: Waiting for the Pods are ready by timeoutSeconds
    PodsReady=false --> Suspended=true(Requeued): Exceeded the timoutSeconds
```

- `replacementTimeoutSeconds`:

```mermaid
stateDiagram-v2
    [*] --> Suspended=false: Workload is admitted
    Suspended=false --> PodsReady=false(1st): Initializing Containers
    PodsReady=false(1st) --> PodsReady=true: Started Coniainers
    PodsReady=true --> PodsReady=false(2nd): Faild Containers
    PodsReady=false(2nd) --> PodsReady=false(2nd): Waiting for replaced Pods are ready by replacementTimeoutSeconds 
    PodsReady=false(2nd) --> Suspended=true(Requeued): Exeeded replacementTimeoutSeconds

```

Author: You understand it just right, thank you for the diagrams

Reviewer 3: I see. In that case, could you clarify the detailed condition when the proposed timeout happens in the KEP?

Author: Added flow chart describing it, similar to yours

We introduce a new workload condition, called `PodsReady`, to indicate
if the workload's startup requirements are satisfied. More precisely, we add

> the condition when `job.status.ready + job.status.uncountedTerimnatedPods + job.status.succeeded` is greater or equal

Reviewer: ```suggestion
the condition when `job.status.ready + len(job.status.uncountedTerimnatedPods.succeeded) + job.status.succeeded` is greater or equal
```
actually

We introduce a new workload condition, called `PodsReady`, to indicate
if the workload's startup requirements are satisfied. More precisely, we add

> the condition when `job.status.ready + len(job.status.uncountedTerimnatedPods.succeeded) + job.status.succeeded` is greater or equal

Reviewer 3: Does this indicate https://github.com/kubernetes-sigs/kueue/blob/6ac33dec0476a452c3406d39b15d9eaff74f44bc/pkg/controller/jobs/job/job_controller.go#L313-L316?

In that case, what if other jobs except batch/v1 Jobs?

Author: I believe the discussion is here, so I'll resolve this one
https://github.com/kubernetes-sigs/kueue/pull/2737#discussion_r1894050196

than `job.spec.parallelism`.

> Note that we count `job.status.uncountedTerminatedPods` - this is meant to prevent flickering of the `PodsReady` condition when pods are transitioning to the `Succeeded` state.

Reviewer 3: Does this improve the existing waitForPodsReady feature even if we do not introduce the recoveryTimeout?

Author: Yes

Reviewer 3: That makes sense.
In that case, are there any improvements related to the flickering issue for other Jobs like RayJob MPIJob?

Author: In case of other types of jobs (including RayJob and MPIJob) we don't rely on counting active pods, but on the job's status:

https://github.com/kubernetes-sigs/kueue/tree/main/pkg/controller/jobs/mpijob#L191-L198

https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobs/rayjob/rayjob_controller.go#L229-L231

Reviewer 3: I know that other job types do not use the active pod count.
My question was, is there any solution to mitigate the flickering of the `PodsReady` condition in other Job types?

Author: IIRC flickering doesn't appear if we don't count the pods

Reviewer 3: That makes sense. Could you expand description that we care the Pods count flickering only for the batch/v1 Job since we do not count active Pods for other Job types and delegate it to third-party Job types operator?

Author: Done

### Timeout on reaching the PodsReady condition

> We introduce two timeouts defined in the `waitForPodsReady.timeoutSeconds` and `waitForPodsReady.recoveryTimeout` fields.

Reviewer 4: ```suggestion
We introduce two timeouts defined in the `waitForPodsReady.timeout` and `waitForPodsReady.recoveryTimeout` fields.
```

First one applies before the job has started. It tracks the time between job getting unsuspended for the first time (the time of unsuspending a job is marked by the Job's
`job.status.startTime` field) and reaching the `PodsReady=true` condition (marked by condition's `.lastTransitionTime`).

```mermaid
flowchart TD;
	start@{ shape: f-circ};
	id1(Suspended=true);
	id2("PodsReady=false
	waits for .timeoutSeconds");
	id3(PodsReady=true);
	id4("Suspended=true (Requeued)");

	start--Workload gets admitted-->id1;
	id1 --> id2;

	id2 --"Doesn't exceed the timeout" --> id3
	id2 --"Exceeds the timeout" --> id4
```

Second one applies when the job has already started and some pod failed while the job is running. It tracks the time between changing `PodsReady` condition to `false` and reaching the

> `PodsReady=true` condition once again.

Reviewer 3: What if the Job is re-failed after the Job gets ready after the recovery?

I meant, what if the Job falls down the below loop?

```mermaid
flowchart TD;
	id3(PodsReady=true);
	id4("PodsReady=false(2nd)
	waits for
	.recoveryTimeout");
	id3 --"Pod failed"--> id4
	id4 --"Pod recovered"--> id3
```

Author: We allow multiple recoveries if that within the timeout

Reviewer 3: I'm worried about falling into the infinite loop, and then the Pods accidentally dominate the resources.
Should we introduce the mitigations for the loop?

We probably want to have a comprehensive timeout separate from `RecoveryTimeout` so that we can evict the infinity loop Pods. Is there any idea instead of a comprehensive timeout?

Author: I think introducing the comprehensive timeout would overengineer the feature. Using batch Job a user can define `.spec.backoffLimit` or `.spec.activeDeadlineSeconds` as a mitigation for the loop. Additionally, without the comprehensive timeout the feature is still not regression

Reviewer 3: > Using batch Job a user can define .spec.backoffLimit or .spec.activeDeadlineSeconds as a mitigation for the loop.

This solution problem is a different persona between Job users and cluster administrators.
I can easily imagine the situations where the cluster admins want to prevent loop Jobs. 

For example, the Job perform downloading dataset or foundation models during preprocessing phase (Pod is running), and then Job failed to perform the preprocesses due to credentials or connection timeout (Pod is failed). 
After that, the Job Pod is recreated based on the failure criteria (Pod is running)

This often could happen within recovery timeout, which means this Job faills to waitForPodsReady loop until the failure cont exceeds to backoffLimit.

Author: My concern is that users can have different types of jobs that will rely on the same config. E.g. a job with 10 000 pods will more likely hit such a timeout compared to smaller job, even if it wouldn't be in an infinite loop. Additionally to track time spend on recovering for a single workload we would need to frequently update its status, as failures with quick recovery can happen very often for large jobs

Reviewer 3: > a job with 10 000 pods will more likely hit such a timeout compared to smaller job, even if it wouldn't be in an infinite loop.

If we completely avoid the problems, we need to entirely remove the cluster-scoped timeout mechanism. But I do not think so since cluster admins tune the timeout parameters based on the workloads specifications.
I meant your proposal could have the same problems.

> Additionally to track time spend on recovering for a single workload we would need to frequently update its status

I could not imagine this. We can just compare the Admitted time and PodsReady time, IIUC.

Here is problem, the infinity loop Job could easily break quota management and unfairly occupy cluster resources. If cluster admins want to allow capacity occupied by an infinity loop Job, they can specify the infinity timeout (nil) as a comprehensive timeout.

Author: > I meant your proposal could have the same problems.

Not really, as the most common scenario would be for one pod to break, so it doesn't matter if it's 1 out of 10 or 1 out of 10 000, it should take more or less the same amount of time. However, if we start to accumulate the recovery time than the scale matters since it's much more probable that 1 out of 10 000 pod breaks in comparison to 1 out of 10.

> I could not imagine this. We can just compare the Admitted time and PodsReady time, IIUC.

Could you elaborate please? I imagined the accumulated recovery time would count intervals between PodsReady=False and PodsReady=True

Author: After discussion offline we came up with mitigation ideas, and I added dedicated paragraphs in the risk and mitigations section regrading this issue.

```mermaid
flowchart TD;
	start@{ shape: f-circ};
	id1(Suspended=true);
	id2("PodsReady=false(1st)");
	id3(PodsReady=true);
	id4("PodsReady=false(2nd)
	waits for
	.recoveryTimeout");
	id5("Suspended=true (Requeued)");

	start--Workload gets admitted-->id1;
	id1 --> id2;

	id2 --"Job started" --> id3
	id3 --"Pod failed"--> id4
	id4 --"Pod recovered"--> id3
	id4 --"timeout	exceeded"--> id5
```

> We introduce new `WorkloadWaitForPodsStart` and `WorkloadWaitForPodsRecovery` reasons to distinguish the reasons of setting the `PodsReady=false` condition.

Reviewer 3: How can existing users migrate the previous "PodsReady" reason to the new reason?
Is there any migration plans?

Author: I believe we can introduce new conditions right away and then continue supporting the old reason (maybe until 0.11) to avoid disruption of running workloads

Reviewer 3: That makes sense.
Could you mention that we treat the old reason the same as `WorkloadWaitForPodsStart`?

Author: Do we want to track historical changes in a KEP?

Reviewer 3: Yes, I think tracking historical reasons as documentation would be helpful.

Author: Done

> We introduce new `WorkloadWaitForPodsReadyStart` and `WorkloadWaitForPodsReadyRecovery` reasons to distinguish the reasons of setting the `PodsReady=false` condition.

Reviewer 4: Maybe `WorkloadWaitForPodsStart` or  and `WorkloadWaitForPodsRecovery`?

Author: Done

```mermaid
flowchart TD;
	start@{ shape: f-circ};
	id1(Suspended=true);
	id2("PodsReady=false
	waits for .timeoutSeconds");
	id3(PodsReady=true);
	id4("Suspended=true (Requeued)");

	start--Workload gets admitted-->id1;
	id1 --> id2;

	id2 --"Doesn't exceed the timeout" --> id3
	id2 --"Exceeds the timeout" --> id4
```

Second one applies when the job has already started and some pod failed while the job is running. It tracks the time between changing `PodsReady` condition to `false` and reaching the
`PodsReady=true` condition once again.

```mermaid
flowchart TD;
	start@{ shape: f-circ};
	id1(Suspended=true);
	id2("PodsReady=false(1st)");
	id3(PodsReady=true);
	id4("PodsReady=false(2nd)
	waits for
	.recoveryTimeout");
	id5("Suspended=true (Requeued)");

	start--Workload gets admitted-->id1;
	id1 --> id2;

	id2 --"Job started" --> id3
	id3 --"Pod failed"--> id4
	id4 --"Pod recovered"--> id3
	id4 --"timeout	exceeded"--> id5
```

We introduce new `WorkloadWaitForPodsStart` and `WorkloadWaitForPodsRecovery` reasons to distinguish the reasons of setting the `PodsReady=false` condition.

> `WorkloadWaitForPodsStart` will be set before the job started and is replacement for the old `PodsReady` reason. 
> `WorkloadWaitForPodsRecovery` will be set after the job started.

Reviewer 3: ```suggestion
- `WorkloadWaitForPodsStart`: This will be set before the job starts. The `PodsReady` reason is used for the older Kueue (< v0.11), is compatible with this new reason, and is handled as equality.
- `WorkloadWaitForPodsRecovery`: This will be set after the job starts.
```

Let us describe the relationship between new and old reasons more clarify.
